### PR TITLE
DLPack interface: fix excpetion handling and improve overheads

### LINKED ISSUE #84056: Handle python exceptions correctly in DLPack Capsule Destructor
### 🐛 Describe the bug

cython code like the following will fail when invoking the python `test` function, with a quite difficult to debug error:

```
import torch
cimport cython
cimport cpython

@cython.no_gc
cdef class DLWrapper:
    cdef object capsule

    def __cinit__(self, tensor):
        self.capsule = tensor.__dlpack__()
        # let's consume it !
        cpython.PyCapsule_SetName(self.capsule, "used_dltensor")

    def __dealloc__(self):
        print("just deallocating some stuff")

cdef inline DLWrapper make_wrapper(tensor):
    cdef DLWrapper dl_wrap = DLWrapper.__new__(DLWrapper, tensor)
    if tensor.numel() == 2:
        raise ValueError("Oh no no no, you better don't do this!")
    return dl_wrap

def test():
    t1 = torch.randn(3)
    dl_wrap_fine = make_wrapper(t1)
    print("phew")
    dl_wrap_bad = make_wrapper(torch.randn(2))
    print("no one was here")
```

This is because here : https://github.com/pytorch/pytorch/blob/c22b8a42e6038ed2f6a161114cf3d8faac3f6e9a/torch/csrc/Module.cpp#L365 if we already have an exception set and the capsule was consumed, we just clear the state out from the outer function.
Instead, one should call `PyErr_Fetch` at the very beginning of that function then `PyErr_Restore` at the end, or something similar.
This seems to be a general misunderstanding of how Python exceptions are handled in Python C, so it may be worth it looking at other bindings whether they have similar issues.

### Versions

Collecting environment information...
PyTorch version: 1.12.1+cpu
Is debug build: False
CUDA used to build PyTorch: Could not collect
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.4 LTS (x86_64)
GCC version: (Ubuntu 9.4.0-1ubuntu1~20.04.1) 9.4.0
Clang version: Could not collect
CMake version: version 3.16.3
Libc version: glibc-2.31

Python version: 3.8.10 (default, Jun 22 2022, 20:18:18)  [GCC 9.4.0] (64-bit runtime)
Python platform: Linux-4.15.0-156-generic-x86_64-with-glibc2.29
Is CUDA available: False
CUDA runtime version: Could not collect
GPU models and configuration: GPU 0: NVIDIA A10
Nvidia driver version: 515.65.01
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] numpy==1.23.2
[pip3] torch==1.12.1+cpu
[pip3] torchaudio==0.12.1+cpu
[pip3] torchvision==0.13.1+cpu
[conda] Could not collect

### PR DESCRIPTION
**Pull request by MatthiasKohl**
Fixes #84056 .

Also improves overheads: the early-out from DLPack's capsule destructor means in most cases, we don't need to set a python error string (within `PyCapsule_GetPointer`), which is costly.

### CODE REVIEW

**Comment by vadimkantorov** [comment]
Maybe related: https://github.com/pytorch/pytorch/issues/43166

it would be good if deleter was not called if it's nullptr / unset

**Comment by MatthiasKohl** [comment]
@vadimkantorov I generally agree with you.
In `fromDLPack` we should check whether the source tensor actually has a `deleter` : we don't know what was the producer of the DLPack tensor.
However, in the capsule destructor that this PR changes, we don't need to check for this since we know the producer of the DLPack tensor: it has to be ATen since otherwise we would not even call this destructor. ATen never sets the deleter to `nullptr` so we don't need this check.

I'll propose the change to `fromDLPack` in PR #84612 which will supersede this PR.

File: torch/_tensor.py [hunk 1]
```diff
@@ -1311,9 +1311,10 @@ def __dlpack__(self, stream=None):
             if self.device.type == "cuda":
                 stream = torch.cuda.ExternalStream(stream)
                 # Only synchronize on different streams
-                if stream != torch.cuda.current_stream:
+                sync_stream = torch.cuda.current_stream()
```
Comment (mruberry) [inline torch/_tensor.py:1314]: Nice catch

**Comment by mruberry** [comment]
Hey @MatthiasKohl! This looks cool. I'm not super familiar with DLPack although I know a little. cc @rgommers if he'd like to take a look, too.

You'll have to sign the PyTorch CLA. 

Is there a test this could add that would have failed before this PR and will succeed after? Maybe one that would exploit the incorrect synchronization?

**Comment by MatthiasKohl** [comment]
@mruberry : This PR should actually be closed, it was superseded by #84612 because of CLA (which goes through Nvidia in this case).

The changes from this PR really shouldn't change any functionality, but will improve performance:
1. for `DLPack_Capsule_Destructor`, the main thing is that it will not "throw an exception" (python C doesn't really throw exceptions but sets an exception state) in the normal case where the consumer of the capsule has already deleted it. Setting this exception state in python C is actually quite expensive and we can easily avoid it.
2. For the stream check, it didn't really matter either since the `if` condition afterwards would have simply always been `False`, which means that we would have always synced with the `current_stream` which was called correctly within the statements of that branch. So this only matters if the user already provided the `current_stream` to `__dlpack__` in which case we do *not* want to sync.

**Pull request closed without merging**